File: steit-csharp/src/Codec/ByteReader.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Steit.Codec {
    public sealed class ByteReader : IReader {
        private readonly IReadOnlyList<Byte> bytes;
        private int offset;

        // public ByteReader(byte[]? bytes = null) : this(Array.AsReadOnly(bytes ?? new byte[0])) { }
        public ByteReader(byte[] bytes = null) : this(Array.AsReadOnly(bytes ?? new byte[0])) { }

        public ByteReader(IReadOnlyList<Byte> bytes) {
            this.bytes = bytes;
            this.offset = 0;
        }

        public int Remaining() {
            return this.bytes.Count - this.offset;
        }

        public byte Read() {
            if (this.Remaining() <= 0) {
                throw new EndOfStreamException();
            }

            return this.bytes[this.offset++];
        }

        public byte[] Read(int count) {
            if (this.Remaining() < count) {
                throw new EndOfStreamException();
            }

            var bytes = new byte[count];

            for (var i = 0; i < count; i++) {
                bytes[i] = this.bytes[this.offset++];
            }

            return bytes;
        }

        public void Skip(int count) {
            if (this.Remaining() < count) {
                throw new EndOfStreamException();
            }

            this.offset += count;
        }
    }
}

File: steit-go/pkg/reader/byte_reader.go
package reader

var _ IReader = (*ByteReader)(nil)

type ByteReader struct {
	bytes  []byte
	offset int
}

func NewByteReader(bytes []byte) IReader {
	return &ByteReader{bytes: bytes, offset: 0}
}

func (b *ByteReader) Remaining() int {
	return len(b.bytes) - b.offset
}

func (b *ByteReader) ReadUint8() byte {
	if b.Remaining() <= 0 {
		panic("end of stream")
	}

	n := b.bytes[b.offset]
	b.offset++

	return n
}

func (b *ByteReader) Read(count int) []byte {
	if b.Remaining() <= 0 {
		panic("end of stream")
	}

	bytes := make([]byte, count)

	for i := 0; i < count; i++ {
		bytes[i] = b.bytes[b.offset]
		b.offset++
	}

	return bytes
}

func (b *ByteReader) Skip(count int) {
	if b.Remaining() < count {
		panic("end of stream")
	}
	b.offset += count
}
